File: src/sql/mysql/protocol/Signature.zig
const Signature = @This();
fields: []Param = &.{},
name: []const u8 = "",
query: []const u8 = "",

pub fn empty() Signature {
    return Signature{
        .fields = &.{},
        .name = "",
        .query = "",
    };
}

pub fn deinit(this: *Signature) void {
    if (this.fields.len > 0) {
        bun.default_allocator.free(this.fields);
    }
    if (this.name.len > 0) {
        bun.default_allocator.free(this.name);
    }
    if (this.query.len > 0) {
        bun.default_allocator.free(this.query);
    }
}

pub fn hash(this: *const Signature) u64 {
    var hasher = std.hash.Wyhash.init(0);
    hasher.update(this.name);
    hasher.update(std.mem.sliceAsBytes(this.fields));
    return hasher.final();
}

pub fn generate(globalObject: *jsc.JSGlobalObject, query: []const u8, array_value: JSValue, columns: JSValue) !Signature {
    var fields = std.array_list.Managed(Param).init(bun.default_allocator);
    var name = try std.array_list.Managed(u8).initCapacity(bun.default_allocator, query.len);

    name.appendSliceAssumeCapacity(query);

    errdefer {
        fields.deinit();
        name.deinit();
    }

    var iter = try QueryBindingIterator.init(array_value, columns, globalObject);

    while (try iter.next()) |value| {
        if (value.isEmptyOrUndefinedOrNull()) {
            // Allow MySQL to decide the type
            try fields.append(.{ .type = .MYSQL_TYPE_NULL, .flags = .{} });
            try name.appendSlice(".null");
            continue;
        }
        var unsigned = false;
        const tag = try types.FieldType.fromJS(globalObject, value, &unsigned);
        if (unsigned) {
            // 128 is more than enought right now
            var tag_name_buf = [_]u8{0} ** 128;
            try name.appendSlice(std.fmt.bufPrint(tag_name_buf[0..], "U{s}", .{@tagName(tag)}) catch @tagName(tag));
        } else {
            try name.appendSlice(@tagName(tag));
        }
        // TODO: add flags if necessary right now the only relevant would be unsigned but is JS and is never unsigned
        try fields.append(.{ .type = tag, .flags = .{ .UNSIGNED = unsigned } });
    }

    if (iter.anyFailed()) {
        return error.InvalidQueryBinding;
    }

    return Signature{
        .name = name.items,
        .fields = fields.items,
        .query = try bun.default_allocator.dupe(u8, query),
    };
}

const bun = @import("bun");
const std = @import("std");
const Param = @import("../MySQLStatement.zig").Param;
const QueryBindingIterator = @import("../../shared/QueryBindingIterator.zig").QueryBindingIterator;

const types = @import("../MySQLTypes.zig");
const FieldType = types.FieldType;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/mysql/protocol/Signature.rs
use crate::jsc::{JSGlobalObject, JSValue};
use bun_sql::mysql::mysql_types::FieldType;
use bun_sql::mysql::protocol::column_definition41::ColumnFlags;

use crate::mysql::my_sql_statement::Param;

#[derive(Default)]
pub struct Signature {
    pub fields: Box<[Param]>,
    pub name: Box<[u8]>,
    pub query: Box<[u8]>,
}

impl Signature {
    pub fn empty() -> Signature {
        Signature {
            fields: Box::default(),
            name: Box::default(),
            query: Box::default(),
        }
    }

    // `deinit` deleted — body only freed owned slices; `Box<[T]>` fields drop automatically.

    // Errors are collapsed into the crate-wide `crate::Error` currency.
    pub fn generate(
        global_object: &JSGlobalObject,
        query: &[u8],
        array_value: JSValue,
        columns: JSValue,
    ) -> crate::Result<Signature> {
        use crate::jsc::js_error_to_mysql;
        use crate::shared::query_binding_iterator::QueryBindingIterator;

        let mut fields: Vec<Param> = Vec::new();
        let mut name: Vec<u8> = Vec::with_capacity(query.len());

        name.extend_from_slice(query);

        // errdefer { fields.deinit(); name.deinit(); } — deleted: `Vec` drops on `?`.

        let mut iter = QueryBindingIterator::init(array_value, columns, global_object)
            .map_err(js_error_to_mysql)?;

        while let Some(value) = iter.next().map_err(js_error_to_mysql)? {
            if value.is_empty_or_undefined_or_null() {
                // Allow MySQL to decide the type
                fields.push(Param {
                    r#type: FieldType::MYSQL_TYPE_NULL,
                    flags: ColumnFlags::empty(),
                });
                name.extend_from_slice(b".null");
                continue;
            }
            let mut unsigned = false;
            let tag =
                crate::mysql::my_sql_value::field_type_from_js(global_object, value, &mut unsigned)
                    .map_err(js_error_to_mysql)?;
            if unsigned {
                name.push(b'U');
                name.extend_from_slice(<&'static str>::from(tag).as_bytes());
            } else {
                name.extend_from_slice(<&'static str>::from(tag).as_bytes());
            }
            // TODO: add flags if necessary right now the only relevant would be unsigned but is JS and is never unsigned
            fields.push(Param {
                r#type: tag,
                flags: if unsigned {
                    ColumnFlags::UNSIGNED
                } else {
                    ColumnFlags::empty()
                },
            });
        }

        if iter.any_failed() {
            return Err(crate::Error::InvalidQueryBinding);
        }

        Ok(Signature {
            name: name.into_boxed_slice(),
            fields: fields.into_boxed_slice(),
            query: Box::<[u8]>::from(query),
        })
    }
}
